File: resources/android/UrlLauncherFunctions.kt
package com.nativephp.plugins.url_launcher

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.content.ActivityNotFoundException
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse

object UrlLauncherFunctions {

    class Execute(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val url = parameters["url"] as? String ?: return BridgeResponse.error(com.nativephp.mobile.bridge.BridgeError.InvalidParameters("No URL provided"))
            val action = parameters["action"] as? String ?: "launch"

            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))

            if (action == "canLaunch") {
                val packageManager = activity.packageManager
                val resolves = intent.resolveActivity(packageManager) != null
                return BridgeResponse.success(mapOf("success" to resolves, "message" to (if (resolves) "Can launch" else "Cannot resolve intent")))
            }

            try {
                activity.startActivity(intent)

                return BridgeResponse.success(mapOf(
                    "success" to true,
                    "message" to "Launched successfully"
                ))
            } catch (e: ActivityNotFoundException) {
                return BridgeResponse.success(mapOf(
                    "success" to false,
                    "message" to e.localizedMessage
                ))
            } catch (e: Exception) {
                return BridgeResponse.error(com.nativephp.mobile.bridge.BridgeError.ExecutionFailed(e.localizedMessage ?: "Unknown error occurred"))
            }
        }
    }

    class GetStatus(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return BridgeResponse.success(mapOf(
                "status" to "ready",
                "version" to "1.0.0"
            ))
        }
    }
}

File: resources/ios/UrlLauncherFunctions.swift
import Foundation
import UIKit

enum UrlLauncherFunctions {

    class Execute: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let urlString = parameters["url"] as? String,
                  let url = URL(string: urlString) else {
                return BridgeResponse.error(message: "No valid URL provided")
            }
            
            let action = parameters["action"] as? String ?? "launch"

            if action == "canLaunch" {
                let canOpen = UIApplication.shared.canOpenURL(url)
                return BridgeResponse.success(data: [
                    "success": canOpen,
                    "message": canOpen ? "Can launch" : "Cannot open URL scheme"
                ])
            }

            // We must dispatch to main thread for UI operations
            var launched = false
            let semaphore = DispatchSemaphore(value: 0)

            DispatchQueue.main.async {
                if UIApplication.shared.canOpenURL(url) {
                    UIApplication.shared.open(url, options: [:]) { success in
                        launched = success
                        semaphore.signal()
                    }
                } else {
                    semaphore.signal()
                }
            }

            semaphore.wait()

            if launched {
                return BridgeResponse.success(data: [
                    "success": true,
                    "message": "Launched successfully"
                ])
            } else {
                return BridgeResponse.success(data: [
                    "success": false,
                    "message": "Could not launch URL"
                ])
            }
        }
    }

    class GetStatus: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            return BridgeResponse.success(data: [
                "status": "ready",
                "version": "1.0.0"
            ])
        }
    }
}